text-spacing: text-spacing-trim: Implement derived half-width font for CoreText platform

vitorroriz opened the pull request:
#### 58c01e33d993c0522a6bd711dd21fb105959857d
<pre>
text-spacing: text-spacing-trim: Implement derived half-width font for CoreText platform
<a href="https://bugs.webkit.org/show_bug.cgi?id=281340">https://bugs.webkit.org/show_bug.cgi?id=281340</a>
<a href="https://rdar.apple.com/137770955">rdar://137770955</a>

Reviewed by Sammy Gill.

This is a preparation patch for text-spacing-trim.
According to specification, trimming should be done
by using OpenType &quot;halt&quot; feature.

This patch creates a derived half-width font by
enabling the OpenType &quot;halt&quot; feature of an existent font.

It also implements a way to verify if a given font
supports such feature.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::halfWidthFont const):
(WebCore::Font::createHalfWidthFont const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::supportsOpenTypeAlternateHalfWidths const):
(WebCore::supportsOpenTypeFeature):
(WebCore::Font::platformCreateHalfWidthFont const):
* Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
(WebCore::Font::platformCreateHalfWidthFont const):
* Source/WebCore/platform/graphics/skia/FontSkia.cpp:
(WebCore::Font::platformCreateHalfWidthFont const):
* Source/WebCore/platform/graphics/win/SimpleFontDataWin.cpp:
(WebCore::Font::platformCreateHalfWidthFont const):

Canonical link: <a href="https://commits.webkit.org/285219@main">https://commits.webkit.org/285219@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7366b3985359ba188cbc8595e30749443fe422b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71882 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51302 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24667 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23091 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73997 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59103 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22911 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/15244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74949 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46535 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61926 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43198 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19396 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21436 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19760 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77722 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16122 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16168 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61949 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64475 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15887 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12647 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47100 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/48169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/49456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47913 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -315,6 +315,13 @@ static void injectTrueTypeCoverage(int type, int selector, CTFontRef font, BitVe
     unionBitVectors(result, source.get());
 }
 
+bool Font::supportsOpenTypeAlternateHalfWidths() const
+{
+    if (m_supportsOpenTypeAlternateHalfWidths == SupportsFeature::Unknown)
+        m_supportsOpenTypeAlternateHalfWidths = supportsOpenTypeFeature(getCTFont(), CFSTR("halt")) ? SupportsFeature::Yes : SupportsFeature::No;
+    return m_supportsOpenTypeAlternateHalfWidths == SupportsFeature::Yes;
+}
```
This should probably follow the pattern in `Font::supportsSmallCaps()`, which ensures that the font actually has glyphs for the feature.

fantasai replied:
Nevermind, Ned pointed out this is usually a GPOS feature so there wouldn't be any alternate glyphs.

fantasai commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -315,6 +315,13 @@ static void injectTrueTypeCoverage(int type, int selector, CTFontRef font, BitVe
     unionBitVectors(result, source.get());
 }
 
+bool Font::supportsOpenTypeAlternateHalfWidths() const
+{
+    if (m_supportsOpenTypeAlternateHalfWidths == SupportsFeature::Unknown)
+        m_supportsOpenTypeAlternateHalfWidths = supportsOpenTypeFeature(getCTFont(), CFSTR("halt")) ? SupportsFeature::Yes : SupportsFeature::No;
+    return m_supportsOpenTypeAlternateHalfWidths == SupportsFeature::Yes;
+}
```
I guess one question would be, should we be checking if it's a listed feature or should we be testing a glyph (e.g. wide brackets) to see if it has an effect? Is it likely to have fonts that would claim to support it, but not have usable support, e.g. because they're working off of a template that injects empty gpos tables or something?

fantasai reviewed (commented):
r=fantasai ; I think it would be interesting to compare this approach to whatever we're doing for `vert` for upright vertical text (which is also a variant that we create), but I simply cannot find that code to make the comparison. But regardless this seems to look correct afaict.

sammygill commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
+        if (!openTypeTag)
+            continue;
+        if (CFStringCompare(openTypeTag, featureTag, 0) == kCFCompareEqualTo)
+            return true;
```
Probably should be indented once to match the rest of the code in the loop?

vitorroriz committed:
text-spacing: text-spacing-trim: Implement derived half-width font for CoreText platform
https://bugs.webkit.org/show_bug.cgi?id=281340
rdar://137770955

Reviewed by Sammy Gill.

This is a preparation patch for text-spacing-trim.
According to specification, trimming should be done
by using OpenType "halt" feature.

This patch creates a derived half-width font by
enabling the OpenType "halt" feature of an existent font.

It also implements a way to verify if a given font
supports such feature.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::halfWidthFont const):
(WebCore::Font::createHalfWidthFont const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::supportsOpenTypeAlternateHalfWidths const):
(WebCore::supportsOpenTypeFeature):
(WebCore::Font::platformCreateHalfWidthFont const):
* Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
(WebCore::Font::platformCreateHalfWidthFont const):
* Source/WebCore/platform/graphics/skia/FontSkia.cpp:
(WebCore::Font::platformCreateHalfWidthFont const):
* Source/WebCore/platform/graphics/win/SimpleFontDataWin.cpp:
(WebCore::Font::platformCreateHalfWidthFont const):

Canonical link: https://commits.webkit.org/285219@main
